---
title: "talk04 练习与作业"
documentclass: ctexart
output:
  rticles::ctex:
    fig_caption: yes
    number_sections: yes
    toc: true
    toc_depth: 2
  word_document: default
  html_document:
    df_print: paged
knit: (
  function(inputFile, encoding) { 

    pSubTitle <- 'GanZiquan-U201912641-talk04-homework'

    rmarkdown::render( 
      input       = inputFile, 
      encoding    = encoding, 
      params      = list(sub_title = pSubTitle),      
      output_file = pSubTitle) })
---

## 练习和作业说明

将相关代码填写入以 \`\`\`{r} \`\`\` 标志的代码框中，运行并看到正确的结果；

完成后，用工具栏里的"Knit"按键生成PDF文档；

**将PDF文档**改为：`姓名-学号-talk04作业.pdf`，并提交到老师指定的平台/钉群。

## Talk04 内容回顾

待写 ...

## 练习与作业：用户验证

请运行以下命令，验证你的用户名。

**如你当前用户名不能体现你的真实姓名，请改为拼音后再运行本作业！**

```{r}
Sys.info()[["user"]]
Sys.getenv("HOME")
```

## 练习与作业1：R session管理

------------------------------------------------------------------------

### **完成以下操作**

-   定义一些变量（比如 x, y , z 并赋值；内容随意）
-   从外部文件装入一些数据（可自行创建一个4行5列的数据，内容随意）
-   保存workspace 到.RData
-   列出当前工作空间内的所有变量
-   删除当前工作空间内所有变量
-   从.RData文件恢复保存的数据
-   再次列出当前工作空间内的所有变量，以确认变量已恢复
-   随机删除两个变量
-   再次列出当前工作空间内的所有变量

```{r}
## 代码写这里，并运行；
#定义一些变量（比如 x, y , z 并赋值；内容随意
rm(list=ls())
x <- c("single", "married", "married", "single");
y <- c(10,100,1000, 10000);
Z <- LETTERS[1:12];

#从外部文件装入一些数据（可自行创建一个4行5列的数据，内容随意）
w=read.table(file="data/Table0.txt")

```

```{r}

#保存workspace 到.RData
save.image(file = "data/Table0.RData");

#列出当前工作空间内的所有变量
ls();

#删除当前工作空间内所有变量
rm(list=ls());

#从.RData文件恢复保存的数据
load(file = "data/Table0.RData");

#再次列出当前工作空间内的所有变量
ls();
```

## 练习与作业2：Factor基础

------------------------------------------------------------------------

### **factor增加**

-   创建一个变量：

`x <- c("single", "married", "married", "single");`

-   为其增加 两个 levels，`single`, `married`;

-   以下操作能成功吗？

`x[3] <- "widowed"`;

-   如果不，请提供解决方案；

```{r}
## 代码写这里，并运行；
x <- c("single", "married", "married", "single");
x <- as.factor(x);
levels(x) <- c("single","married");
#x[3] <- "widowed";
#不行，因为x为factor，只允许接受single和married
levels(x) <- c(levels(x), "widowed");
x[ length(x) + 1 ] <- "widowed";
x[3] <- "widowed";
x;
```

------------------------------------------------------------------------

### **利用factor排序**

以下变量包含了几个月份，请使用`factor`，使其能按月份，而不是英文字符串排序：

`mon <- c("Mar","Nov","Mar","Aug","Sep","Jun","Nov","Nov","Oct","Jun","May","Sep","Dec","Jul","Nov");`

```{r}
## 代码写这里，并运行；
mon <- c("Mar","Nov","Mar","Aug","Sep","Jun","Nov","Nov",
         "Oct","Jun","May","Sep","Dec","Jul","Nov");
month_levels <- c("Jan", "Feb", "Mar", "Apr", "May","Jun",
                  "Jul", "Aug", "Sep", "Oct", "Nov", "Dec");
x1 <- factor(mon, levels = month_levels);
sort(x1);
```

------------------------------------------------------------------------

### **forcats的问题**

`forcats`包中的 fct_inorder, fct_infreq和 fct_inseq 函数的作用是什么？

This family of functions changes only the order of the levels

fct_inorder : by the order in which they first appear.

fct_infreq : by number of observations with each level( large first)

fct_inseq : by numeric value of level

请使用`forcats`包中的`gss_cat`数据举例说明

```{r}
## 代码写这里，并运行；
if (!require("forcats")){ 
  chooseCRANmirror();
  install.packages("forcats",destdir = "D:/resourse/software/Rproject4.1.1/download packages");
} 
library("forcats");
head(gss_cat);
attach(gss_cat);
head(fct_inorder(marital),n=20)
head(fct_infreq(rincome),n=30)

f<-factor(1:6,levels=c("1 ","2","3","4","5","6"))
fct_inseq(f)

```

## 练习与作业3：用 mouse genes 数据做图

------------------------------------------------------------------------

### **画图**

1.  用 readr 包中的函数读取 mouse genes 文件 （从本课程的 Github 页面下载 data/talk04/ ）
2.  选取常染色体的基因
3.  画以下两个基因长度 boxplot :

-   按染色体序号排列，比如 1, 2, 3 .... X, Y
-   按基因长度 中值 排列，从 短 -\> 长 ...

```{r}
## 代码写这里，并运行；
library(readr)
library(ggplot2)
mouse.tibble <- read_delim( file = "data/talk04/mouse_genes_biomart_sep2018.txt",
delim = "\t", quote = "" )
mouse.tibble.chr <-
mouse.tibble %>% filter( `Chromosome.scaffold.name` %in% c( "1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","X","Y" ) );
plot1 <- 
ggplot( data = mouse.tibble.chr, 
aes( x = reorder( `Chromosome/scaffold name`, 
-as.integer(mouse.tibble.chr$`Chromosome/scaffold name`)),
y = `Transcript length (including UTRs and CDS)` ) ) +
geom_boxplot() + 
coord_flip() + 
ylim( 0, 5000 )
plot1
plot2 <-
ggplot( data = mouse.tibble.chr,
aes( x = reorder( `Chromosome/scaffold name`,
-`Transcript length (including UTRs and CDS)`,
median ),
y = `Transcript length (including UTRs and CDS)` ) ) +
geom_boxplot() +
coord_flip() +
ylim( 0, 5000 ) 
plot2
```
